### PR TITLE
[Merged by Bors] - activation: panic if post service exits unexpectedly

### LINKED ISSUE #5267: Node should panic in PoST supervised mode if service doesn't connect
## Description

There are a few ways in which a node can be configured to start in PoST supervised mode but fail to start smeshing. The most common ones are:

- `grpc-private-services`/`grpc-tls-services` are missing `post` handler and therefor do not accept connections from the post service
- `node-address` is not the correct address, i.e. it doesn't match the host + port defined by `grpc-private-listener`/`grpc-tls-listener`

This will result in the post service starting up and trying to connect every 5 seconds, but failing to do so.

With https://github.com/spacemeshos/go-spacemesh/pull/5259 a possibly incorrect config is detected and logged by the node during startup. Additionally if the configuration causes the PoST service to be unable to connect to the node, the node should panic and shut down so a misconfigured setup doesn't go unnoticed by the user.

### Acceptance criteria

- [x] spacemeshos/post-rs#151
- update the node supervisor to not try to restart the post service when it exits unexpectedly but instead panic with a message that the node is shutting down due to the post service exiting unexpectedly:
  - since the output of the post service is captured and added to the node logs it will show that connection/configuration problems were the cause

### PR DESCRIPTION
**Pull request by fasmat**
## Motivation
Closes #5267 merge after https://github.com/spacemeshos/post/pull/249

## Changes
- PoST supervisor was updated to add `--max-retries` as cmdline parameter to Post service
- Number of retries can be configured using the `PostSupervisorConfig` (not exposed via node config or cmdline parameters) and defaults to 10. This gives the node about 50 seconds to be ready for the post service to connect after starting it

## Test Plan
- Updated existing tests to match new behavior
- Add test that checks if incorrect setup that causes the post service to be unable to connect to the node causes `zap.Fatal` to be called

## TODO

- [x] Explain motivation or link existing issue(s)
- [x] Test changes and document test plan
- [x] Update documentation as needed
- [x] Update [changelog](../CHANGELOG.md) as needed

### CODE REVIEW

File: activation/post_supervisor.go [hunk 1]
```diff
@@ -232,67 +225,69 @@ func (ps *PostSupervisor) runCmd(
...
-		if cmdCfg.CACert != "" {
-			args = append(args, "--ca-cert", cmdCfg.CACert)
-		}
-		if cmdCfg.Cert != "" {
-			args = append(args, "--cert", cmdCfg.Cert)
-		}
-		if cmdCfg.Key != "" {
-			args = append(args, "--key", cmdCfg.Key)
-		}
+		// "--watch-pid", strconv.Itoa(os.Getpid()),
```
Comment (fasmat) [inline activation/post_supervisor.go]: @poszu I think there is a bug in the Post service: If I use both `--watch-pid` and `--max-retries` the service will print "shutting down" after the specified number of retries, but won't exit.

File: go.mod [hunk 2]
```diff
@@ -39,7 +39,7 @@ require (
 	github.com/spacemeshos/go-scale v1.1.12
 	github.com/spacemeshos/merkle-tree v0.2.3
 	github.com/spacemeshos/poet v0.9.7
-	github.com/spacemeshos/post v0.10.1
+	github.com/spacemeshos/post v0.10.2-rc1
```
Comment (fasmat) [inline go.mod:42]: TODO: Release a new stable version of post before merging.

**Comment by fasmat** [comment]
@poszu the new version of post-rs seems to not compile on windows and shows unusual errors on macOS (it also complains that it was built for a newer macOS than it is executed on).

**Comment by fasmat** [comment]
The new version of post-rs also seems to validate `NumUnits` to be at least 4 which broke a few tests: https://github.com/spacemeshos/go-spacemesh/actions/runs/6973779216/job/18978370093?pr=5300

File: activation/post_supervisor.go (hunk 1 above)
Comment (poszu) [reply activation/post_supervisor.go]: You are right. I fixed it in https://github.com/spacemeshos/post-rs/pull/155.

**Comment by poszu** [comment]
> The new version of post-rs also seems to validate `NumUnits` to be at least 4 which broke a few tests: https://github.com/spacemeshos/go-spacemesh/actions/runs/6973779216/job/18978370093?pr=5300

This is the mainnet default. You can overwrite it with `--min-num-units`.

**Comment by poszu** [comment]
> @poszu the new version of post-rs seems to not compile on windows and shows unusual errors on macOS (it also complains that it was built for a newer macOS than it is executed on).

Weird, it builds on Windows nicely in post-rs CI.

Regarding Mac - it was previously built with Mac SDK 12.3 but is now built with the default SDK (no overwrite). Building with an older SDK was a workaround for a bug in RadnomX lib (causing crashes on Mac) that is now fixed.

**Comment by fasmat** [comment]
bors merge

File: activation/post_supervisor.go [hunk 3]
```diff
@@ -232,67 +225,72 @@ func (ps *PostSupervisor) runCmd(
...
+		events.EmitPostServiceStopped()
+		if err := eg.Wait(); err != nil {
+			ps.logger.Warn("output reading goroutine failed", zap.Error(err))
 		}
-		ps.logger.Warn("post service exited", zap.Error(err))
-		eg.Wait()
+		return nil
 	}
+	eg.Wait()
+	ps.logger.Fatal("post service exited", zap.Error(err))
```
Comment (dshulyak) [inline activation/post_supervisor.go:294]: i think it makes more sense to raise Fatal in node when error will bubble up. there are several examples of this, for example peersync process

**Review by dshulyak — APPROVED** [review]

File: activation/post_supervisor.go (hunk 3 above)
Comment (fasmat) [reply activation/post_supervisor.go:294]: You mean just returning an error here and handle the error with `log.Fatal` in `node.go`? 

The problem is that the error here happens asynchronously and doesn't bubble up at the moment. Returning it here would do nothing (besides preventing the internal `errgroup` from accepting new go routines).

**Comment by fasmat** [comment]
bors merge

**Pull request closed without merging**